libfdt/fdt_ro.c

```diff
@@ -269,6 +269,9 @@ int fdt_path_offset_namelen(const void *fdt, const char *path, int namelen)
 
 		if (!p)
 			return -FDT_ERR_BADPATH;
+		if (!can_assume(VALID_INPUT) && *p != '/')
+			return -FDT_ERR_BADVALUE;
+
```

Comment:
Would it make sense to write:

    /aliases {
        self = "&somelabel";
    }
? With the previous commit it would work I think. I think I would have picked `-EFDT_ERR_BADPATH` as error code.

Reply:
> Would it make sense to write:
> 
> ```
> /aliases {
>     self = "&somelabel";
> }
> ```
> 
> ? With the previous commit it would work I think. 

I can't see the use case. For this to resolve correctly, the dtb would have to have been built with `-@`, and then the `somelabel` property would still contain the absolute path, so there's no way this could even make the dtb smaller. So no, I would consider such a .dts file malformed, and I don't see any reason we should allow anything but absolute paths on the RHS of alias and symbol properties.

> I think I would have picked `-EFDT_ERR_BADPATH` as error code.

I believe my rationale in the commit message makes sense, and I certainly don't want to overload FDT_ERR_BADPATH more when there's such a fitting new error code to use.

Reply:
> For this to resolve correctly, the dtb would have to have been built with `-@`, and then the `somelabel` property would still contain the absolute path, so there's no way this could even make the dtb smaller. So no, I would consider such a .dts file malformed, and I don't see any reason we should allow anything but absolute paths on the RHS of alias and symbol properties.

It can be smaller; with the following dts in test.dts

```dts
/dts-v1/;

/ {
	aliases {
		sl1 = "&somelabel";
		sl2 = &somelabel;
	};

	somelabel: somenodewithalongname {
		someproperty = "somevalue";
	};
};
```

and

```diff
diff --git a/libfdt/fdt_ro.c b/libfdt/fdt_ro.c
index d30c757180b7..c06c5d24d959 100644
--- a/libfdt/fdt_ro.c
+++ b/libfdt/fdt_ro.c
@@ -269,8 +269,10 @@ int fdt_path_offset_namelen(const void *fdt, const char *path, int namelen)
 
                if (!p)
                        return -FDT_ERR_BADPATH;
+#if 0
                if (!can_assume(VALID_INPUT) && *p != '/')
                        return -FDT_ERR_BADVALUE;
+#endif
 
                offset = fdt_path_offset(fdt, p);
 
```

```shell
$ dtc -@ -I dts -O dtb test.dts > test.dtb
test.dts:5.3-22: Warning (alias_paths): /aliases:sl1: aliases property is not a valid node (&somelabel)
$ fdtget test.dtb sl1 someproperty
somevalue
```

So it works fine and the sl1 alias is shorter than the equivalent sl2 one.

Reply:
> > For this to resolve correctly, the dtb would have to have been built with `-@`, and then the `somelabel` property would still contain the absolute path, so there's no way this could even make the dtb smaller. So no, I would consider such a .dts file malformed, and I don't see any reason we should allow anything but absolute paths on the RHS of alias and symbol properties.
> 
> It can be smaller; with the following dts in test.dts
> 

Yes, of course, if the base assumption is that the dts will be built with -@ regardless, so the absolute path will be listed under the __symbols__ node, of course the string `&somelabel` will be shorter than the having the value of `sl1` be another copy of that absolute path.

However, that is entirely besides the point. I still completely fail to see the actual use case, or why we should allow aliases or symbols to resolve to anything but an absolute path, and stand by my "I would consider such a dts malformed". And current dtc obviously agrees with me, per the warning output above.

Reply:
> > I think I would have picked `-EFDT_ERR_BADPATH` as error code.
> 
> I believe my rationale in the commit message makes sense, and I certainly don't want to overload FDT_ERR_BADPATH more when there's such a fitting new error code to use.

Hm, to sum up, with

```
/dts-v1/;

/ {
	aliases {
		sa1 = &somelabel;
		sa2 = "/nonexisting";
		sa3 = "invalid";
	};

	somelabel: somenodewithalongname {
		someproperty = "somevalue";
	};
};
```

we get for `fdtget test.dtb sa$i someproperty` respectively:
 * `i = 1`, success: `"somevalue"`
 * `i = 2`, error `-FDT_ERR_NOTFOUND`
 * `i = 3`, error `-FDT_ERR_BADVALUE`
 * `i = 4`, error `-FDT_ERR_BADPATH`

Looking at it from a user's perspective: The user asks for a property at the path "sa3". Replying with

    FDT_ERR_BADPATH: Function was passed a badly formatted path

seems right to me and also matches the behavior of fdtget before your changes(!). However replying with

    FDT_ERR_BADVALUE: Device tree has a property with an unexpected value.

(while being correct) is a diagnosis about the property `/aliases/sa3` and not about the path "sa3" that the user provided and so only something like an implementation detail.

I'm not authoritative here, so I'll stop arguing here and let @dgibson think and decide about that.

Reply:
I don't think we should extend the semantics of `/aliases`.  That's been around since the ancient history of OF, and I think extending it is of very little use here.  Noting that you can already do:

```
/ {
        aliases {
                sa = &label;
        }
}
```
Where dtc will expand that to the full path.

I'm not 100% sure, but I believe it's also valid for aliases to reference other aliases.  I couldn't quickly find absolute confirmation in IEEE1275, but I think this was valid an used in practice in the old OF days.  For example:
```
/ {
        aliases {
                sa1 = "/foo/bar";
                sa2 = "sa1/baz";
        };
}
````

Reply:
> > I'm not 100% sure, but I believe it's also valid for aliases to reference other aliases. I couldn't quickly find absolute confirmation in IEEE1275, but I think this was valid and used in practice in the old OF days. For example:
> > ```
> > / {
> >         aliases {
> >                 sa1 = "/foo/bar";
> >                 sa2 = "sa1/baz";
> >         };
> > }
> > ```
> 
> Well, if that's really the case, then of course my sanity check patch should be removed. 

Ah, but https://www.devicetree.org/specifications/ is quite clear:
```
Each property of the ``/aliases`` node defines an alias. The property name
specifies the alias name. The property value specifies the full path to
a node in the devicetree.
```
and
```
An alias value is a device path and is encoded as a string. The value
represents the full path to a node, but the path does not need to refer
to a leaf node.
```
That text seems to be essentially unchanged since it first appeared in that git repo.

Reply:
> > > I'm not 100% sure, but I believe it's also valid for aliases to reference other aliases. I couldn't quickly find absolute confirmation in IEEE1275, but I think this was valid and used in practice in the old OF days. For example:
> > > ```
> > > / {
> > >         aliases {
> > >                 sa1 = "/foo/bar";
> > >                 sa2 = "sa1/baz";
> > >         };
> > > }
> > > ```
> > 
> > 
> > Well, if that's really the case, then of course my sanity check patch should be removed.
> 
> Ah, but https://www.devicetree.org/specifications/ is quite clear:

Eh.. maybe..
 
> ```
> Each property of the ``/aliases`` node defines an alias. The property name
> specifies the alias name. The property value specifies the full path to
> a node in the devicetree.
> ```
> 
> and
> 
> ```
> An alias value is a device path and is encoded as a string. The value
> represents the full path to a node, but the path does not need to refer
> to a leaf node.
> ```

So, both of these look to be paraphrased from IEEE1275.  The ambiguity, such as it is, is in the meaning of "full path".  1275 uses the term `device-path` in the equivalent place, and I think that term may allow the first component to be an alias, rather than `/`, though I'm having trouble pinning down something which says that one way or another.  I do see `device-path` used in other contexts of 1275 where I'm pretty certain aliases are allowed (e.g. the value of the `bootpath` property of `/chosen`).

> That text seems to be essentially unchanged since it first appeared in that git repo.

As you'd expect for something more or less inherited from 1275.

Reply:
> The ambiguity, such as it is, is in the meaning of "full path". 

Alright, I've pushed a version without that sanity check (and with s/nodename/path/).